### PR TITLE
Add libmode melange

### PR DESCRIPTION
**Pull request by jchavarri**
Second attempt at #1, but this time:
- trying to leave OCaml modules like `Cm_kind` or `Mode` untouched
- avoid sharing `cmi` rules between byte and melange modes
- put melange artifacts into its own obj dir folder `melange`
- rename melange mode to `melange_experimental`

### CODE REVIEW

File: src/dune_lang/lib_mode.ml [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+let to_string = choose Mode.to_string "melange_experimental"
+
+let encode t = Dune_sexp.Encoder.string (to_string t)
+
+let decode =
+  let open Dune_sexp.Decoder in
+  enum
+    [ ("byte", Ocaml Byte)
+    ; ("native", Ocaml Native)
+    ; ("melange_experimental", Melange)
```
Comment (rgrinberg) [inline src/dune_lang/lib_mode.ml:89]: The name `melange` will do just fine.

To make a feature as experimental, we should do the following:
1. Introduce a melange "syntax"
2. Set the version of this syntax as pre 1.0 (0.1 for example).
3. Guard this new melange mode behind the new syntax.

If you need more pointers on how to do this, let me know. The coq rules is a good reference on how to do this.

File: src/dune_rules/install_rules.ml [hunk 2]
```diff
@@ -160,26 +161,32 @@ end = struct
               | Some f -> [ (cm_kind, f) ]
             else []
           in
-          let open Cm_kind in
-          [ if_ true (Cmi, cm_file Cmi)
-          ; if_ native (Cmx, cm_file Cmx)
-          ; if_ (byte && virtual_library) (Cmo, cm_file Cmo)
+          let open Lib_mode.Cm_kind in
+          [ if_ (byte || native) (Ocaml Cmi, cm_file (Ocaml Cmi))
+          ; if_ melange (Melange Cmi, cm_file (Melange Cmi))
```
Comment (rgrinberg) [inline src/dune_rules/install_rules.ml:166]: For now, I would drop the installation of melange artifacts. You have no use for this feature yet, and this change alone isn't enough to make it work (you need symmetrical changes to read this stuff from dune-package).

File: src/dune_rules/lib_file_deps.ml [hunk 3]
```diff
@@ -1,22 +1,33 @@
 open Import
 
 module Group = struct
-  type t =
+  type ocaml =
     | Cmi
     | Cmx
+
+  type melange =
```
Comment (rgrinberg) [inline src/dune_rules/lib_file_deps.ml:8]: There's already a few instances of this type already. What about introducing a `Melange` module (in dune_rules) that will define common types/facts about melange.

File: src/dune_rules/lib_rules.ml [hunk 4]
```diff
@@ -367,12 +368,23 @@ let setup_build_archives (lib : Dune_file.Library.t) ~top_sorted_modules ~cctx
       ~top_sorted_modules ()
   in
   let* () =
-    Mode.Dict.Set.iter_concurrently modes ~f:(fun mode ->
+    Mode.Dict.Set.iter_concurrently modes.ocaml ~f:(fun mode ->
         build_lib lib ~native_archives ~dir ~sctx ~expander ~flags ~mode ~scope
           ~cm_files)
+  and* () =
+    Memo.when_ modes.melange (fun () ->
+        (* Create empty library target, so that melange libraries modules rules get added to build implicitly *)
```
Comment (rgrinberg) [inline src/dune_rules/lib_rules.ml:376]: I'm not sure why you need this.

File: src/ocaml/cm_kind.mli [hunk 5]
```diff
@@ -9,6 +9,8 @@ val compare : t -> t -> Ordering.t
 
 val all : t list
 
+val choose : 'a -> 'a -> 'a -> t -> 'a
```
Comment (rgrinberg) [inline src/ocaml/cm_kind.mli:12]: Is this function used anywhere?

File: src/dune_rules/exe.ml [hunk 6]
```diff
@@ -81,7 +81,10 @@ module Linkage = struct
           | Native -> Native
           | Best ->
             if Result.is_ok ctx.ocamlopt then Native
-            else Byte_with_stubs_statically_linked_in)
+            else Byte_with_stubs_statically_linked_in
+          | Melange ->
+            User_error.raise
+              [ Pp.textf "Linking is not supported in melange mode" ])
```
Comment (rgrinberg) [inline src/dune_rules/exe.ml:87]: The record inside `Other` should use the old `mode` to avoid this check if possible.

File: src/dune_rules/dune_file.ml [hunk 7]
```diff
@@ -403,6 +403,7 @@ module Mode_conf = struct
       | Byte
       | Native
       | Best
+      | Melange
```
Comment (rgrinberg) [inline src/dune_rules/dune_file.ml:406]: I think this is used only for binaries which we aren't supporting yet.

File: src/dune_rules/compilation_context.mli [hunk 8]
```diff
@@ -84,7 +85,9 @@ val package : t -> Package.t option
 
 val vimpl : t -> Vimpl.t option
 
-val modes : t -> Mode.Dict.Set.t
+val modes : t -> Lib_mode.Dict.Set.t
+
+val ocaml_modes : t -> Mode.Dict.Set.t
```
Comment (rgrinberg) [inline src/dune_rules/compilation_context.mli:90]: Do we really need this function? It seems to be a simpler wrapper of `modes`.

**Review by rgrinberg — COMMENTED** [review]
I added a preliminary review. In general, things look much better now and I think we're on the path to upstream this. I didn't review everything, but I left enough comments for you to work on.

Let me know if any of the comments aren't clear enough. We can also go through a review together over a call.

File: src/dune_rules/lib_rules.ml (hunk 4 above)
Comment (jchavarri) [reply src/dune_rules/lib_rules.ml:376]: If I don't add this rule, then no `ml` sources are added as targets of melange libraries, so the rules set remains empty. Is there a way to somehow add a rule that has `obj_deps` as deps but performs no action?

If I try to remove the `write_file` action:

```ml
let obj_deps =
          Action_builder.paths (Cm_files.melange_objects_and_cms cm_files)
        in
        Super_context.add_rule ~dir sctx ~loc:lib.buildable.loc
          (let open Action_builder.With_targets.O in
          Action_builder.with_no_targets obj_deps))
```

I get this error:
```
This expression has type unit Action_builder.With_targets.t
but an expression was expected of type
  Action.Full.t Action_builder.With_targets.t
Type unit is not compatible with type Action.Full.t
```

File: src/dune_rules/compilation_context.mli (hunk 8 above)
Comment (jchavarri) [reply src/dune_rules/compilation_context.mli]: I guess not. Removed in https://github.com/jchavarri/dune/pull/4/commits/190e2b9285f4b344e4d4bf0fe2c40ab7761d22e7.

File: src/dune_rules/lib_rules.ml (hunk 4 above)
Comment (rgrinberg) [reply src/dune_rules/lib_rules.ml:376]: I’m still not certain what’s the root of the issue. build_lib is for building library archives, which melange doesn’t have. So is it just that you need some fake target to build a library with melange?

For example:

(library
 (name foo))

And you don’t want to write:

dune build foo.cmj

File: src/dune_rules/dune_file.ml (hunk 7 above)
Comment (jchavarri) [reply src/dune_rules/dune_file.ml:406]: Agreed. I tried to remove it and it helps a lot removing edge cases where I was raising exns. However, parsing the `modes` stanza fails with 

Seems `Dune_file.parse_file_includes` is called while parsing the dune file? Here's the relevant stack trace (obtained by replacing user_error exn with regular one in `Decode.enum`):

```
Called from Dune_sexp__Decoder.parse in file "src/dune_sexp/decoder.ml", line
  249, characters 13-29
Called from Stdune__List.concat_map.aux in file "otherlibs/stdune/list.ml",
  line 41, characters 15-18
Called from Dune_rules__Dune_file.Stanzas.parse_file_includes in file
  "src/dune_rules/dune_file.ml", line 2177, characters 4-50
Called from Dune_rules__Dune_file.Stanzas.parse in file
  "src/dune_rules/dune_file.ml", line 2190, characters 6-55
Called from Dune_rules__Dune_file.parse in file
  "src/dune_rules/dune_file.ml", line 2239, characters 17-50
Called from Dune_rules__Dune_load.Dune_files.interpret.impl in file
  "src/dune_rules/dune_load.ml", line 202, characters 23-65
Called from Fiber__Scheduler.exec in file "src/fiber/scheduler.ml", line 73,
  characters 8-11
-> required by ("Dune_files.interpret", <opaque>)
```

Comment (rgrinberg) [reply src/dune_rules/dune_file.ml:406]: Never mind, it's the link mode that is used for binaries. This is the right change.

Comment (rgrinberg) [reply src/dune_rules/dune_file.ml:406]: > Here's the relevant stack trace (obtained by replacing user_error exn with regular one in Decode.enum):

You can run dune with `--debug-backtraces` to see the full trace.

File: src/dune_rules/obj_dir.ml [hunk 9]
```diff
@@ -102,26 +106,43 @@ module Local = struct
     ; obj_dir : Path.Build.t
     ; native_dir : Path.Build.t
     ; byte_dir : Path.Build.t
+    ; melange_dir : Path.Build.t
```
Comment (rgrinberg) [inline src/dune_rules/obj_dir.ml:109]: I'm not sure a single directory would be enough in the case where `private_modules` are used.

It would be worth it to add a test that exercises this feature in melange:

```
(library
 (private_modules bar)
 (name foo))
```

File: src/dune_rules/dune_package.ml [hunk 10]
```diff
@@ -86,7 +86,7 @@ module Lib = struct
        ; field_o "default_implementation" (no_loc Lib_name.encode)
            default_implementation
        ; field_o "main_module_name" Module_name.encode main_module_name
-       ; field_l "modes" sexp (Mode.Dict.Set.encode modes)
+       ; field_l "modes" sexp (Lib_mode.Dict.Set.encode modes)
```
Comment (rgrinberg) [inline src/dune_rules/dune_package.ml:89]: I think we should use the old modes here for all the decoders. We don't support installing melange artifacts yet.

File: src/dune_rules/dune_file.ml (hunk 7 above)
Comment (jchavarri) [reply src/dune_rules/dune_file.ml]: > it's the link mode that is used for binaries. This is the right change

The "mode conf" used for libraries and other cases was shared type for all cases. I was able to clean up link mode + all other cases by adding a new module `Mode_conf.Lib` that includes the melange mode, and is only used for libraries.

I am not 100% sure about it as it adds more boilerplate, but code looks cleaner now, and cases that should not be handled at the moment for melange (exes, installation, linking) are gone, please check https://github.com/jchavarri/dune/pull/4/commits/e8c901c43eb07c4931dbc42a587d18f64388b182

Comment (rgrinberg) [reply src/dune_rules/dune_file.ml]: Looks fine to me. I'd take straightforward boilerplate over imprecise types.

File: src/dune_rules/dune_package.ml (hunk 10 above)
Comment (jchavarri) [reply src/dune_rules/dune_package.ml]: Removed in https://github.com/jchavarri/dune/pull/4/commits/9bbdd97b08870aa1048ab1afbc26dddb9b55701b.

File: src/ocaml/cm_kind.mli (hunk 5 above)
Comment (jchavarri) [reply src/ocaml/cm_kind.mli]: nope, I used it at some point but then wasn't needed. removed in https://github.com/jchavarri/dune/pull/4/commits/7854dddcaeaf468a38dcff9be5d4531061ada6fc.

File: src/dune_rules/exe.ml (hunk 6 above)
Comment (jchavarri) [reply src/dune_rules/exe.ml]: This was fixed in https://github.com/jchavarri/dune/commit/e8c901c43eb07c4931dbc42a587d18f64388b182 (see [related discussion](https://github.com/jchavarri/dune/pull/4#discussion_r992833264))

File: src/dune_rules/install_rules.ml (hunk 2 above)
Comment (jchavarri) [reply src/dune_rules/install_rules.ml]: Fixed all this in https://github.com/jchavarri/dune/pull/4/commits/9f91d01c041930185c650e4a5e02cba2fcab8edf.

File: src/dune_rules/lib_rules.ml (hunk 4 above)
Comment (jchavarri) [reply src/dune_rules/lib_rules.ml:376]: What you suggests works, and I can use `.js` instead of `.cmj`. But I have to provide the full path to the `_build` location where the artifact will be placed, e.g:

```
dune build src/.main.objs/melange/main.js
```

works fine.

But if I use the "in-source" path, it won't work:

```
dune build src/main.js
Error: Don't know how to build src/main.js
```

The first option would work if having this hack for .cma is not possible, although it's a bit cumbersome from user perspective.

File: src/dune_rules/lib_file_deps.ml (hunk 3 above)
Comment (jchavarri) [reply src/dune_rules/lib_file_deps.ml]: I attempted this in https://github.com/jchavarri/dune/pull/4/commits/0cf410efbee293c2acc5e103f52cc8119b678875. I had to move `Lib_mode` to `dune_rules` as well, but I believe that should be fine. Also removed some functions from `Lib_mode` that were not used.

I think code looks better now (more similarities between the way OCaml and Melange modes are handled), wdyt?

File: src/dune_lang/lib_mode.ml (hunk 1 above)
Comment (jchavarri) [reply src/dune_lang/lib_mode.ml]: Thanks for the explanation. I applied the changes in https://github.com/jchavarri/dune/pull/4/commits/f0b3448473d5bc2f36ba558ba31d6920c7a764f8. I understand no changes are needed on the decoding side? (i.e. `Mode_conf.Lib.Set.decode`)

File: src/dune_rules/obj_dir.ml (hunk 9 above)
Comment (jchavarri) [reply src/dune_rules/obj_dir.ml:109]: Good to know. As next step, I am working on bringing over the tests from #1, but most probably will do so in a separate PR to not increase the size of this one.

I will include private modules tests in the melange suite 👍

Comment (jchavarri) [reply src/dune_rules/obj_dir.ml:109]: ftr: https://github.com/jchavarri/dune/pull/5

You were right, private_modules are broken. Looking at the code, I am not sure I understand why though. My initial question would be if this is something that would be fixed when installation support is added for melange libs? In which case, might be worth focusing on that? (rather than adding a fix for private modules in uninstalled libs)

File: src/dune_rules/lib_rules.ml (hunk 4 above)
Comment (jchavarri) [reply src/dune_rules/lib_rules.ml]: We can work around with build path for now, removed cma hack in https://github.com/jchavarri/dune/pull/4/commits/7882f6ff3b30aea09008c208a4a3f67d33bb2db8.

File: src/dune_rules/obj_dir.ml (hunk 9 above)
Comment (rgrinberg) [reply src/dune_rules/obj_dir.ml:109]: Installation support would be a separate thing. The reason why it's broken is roughly as follows: private modules must have their cmi's kept in a special directory. This is so that we're able to hide them from other libraries. In the current implementation, you only have one directory for melange cmi's, so you're unable to enforce the separation.

File: src/dune_rules/cm_files.mli [hunk 11]
```diff
@@ -15,8 +15,12 @@ val make :
   -> unit
   -> t
 
+val filter_excluded_modules : t -> Module.t list -> Module.t list
+
 val unsorted_objects_and_cms : t -> mode:Mode.t -> Path.t list
 
+val melange_objects_and_cms : t -> Path.t list
```
Comment (rgrinberg) [inline src/dune_rules/cm_files.mli:22]: Is this function being used anywhere?

File: src/dune_rules/dune_file.mli [hunk 12]
```diff
@@ -117,6 +117,43 @@ module Mode_conf : sig
 
     val eval : t -> has_native:bool -> Mode.Dict.Set.t
   end
+
+  module Lib : sig
+    type mode_conf = t
```
Comment (rgrinberg) [inline src/dune_rules/dune_file.mli:122]: You can do `type mode_conf := t` to avoid this alias from being exported.

File: src/dune_rules/dune_file.mli [hunk 13]
```diff
@@ -117,6 +117,43 @@ module Mode_conf : sig
...
+
+    module Map : sig
+      type nonrec 'a t =
+        { ocaml : 'a Map.t
+        ; melange : 'a
+        }
+    end
+
+    module Set : sig
+      type mode_conf = t
```
Comment (rgrinberg) [inline src/dune_rules/dune_file.mli:140]: ditto

File: src/dune_rules/melange.ml [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    | Cmi -> ".cmi"
+    | Cmj -> ".cmj"
+
+  let to_dyn =
+    let open Dyn in
+    function
+    | Cmi -> variant "cmi" []
+    | Cmj -> variant "cmj" []
+
+  module Dict = struct
```
Comment (rgrinberg) [inline src/dune_rules/melange.ml:24]: Could you rename this `Dict` to `Map`.  I know we don't use this consistently, but we ended up deciding on `Map` in the codebase but didn't yet change all the existing code.

File: src/dune_rules/install_rules.ml [hunk 15]
```diff
@@ -174,12 +177,17 @@ end = struct
...
+            in
+            (cm_kind, cmt)
+          in
           List.concat_map installable_modules.impl ~f:(fun m ->
               common m
-              @ List.filter_map Ml_kind.all ~f:(fun ml_kind ->
-                    let open Option.O in
-                    let+ cmt = Obj_dir.Module.cmt_file obj_dir m ~ml_kind in
-                    (Cm_kind.Cmi, cmt))
+              @ List.filter_map Ml_kind.all ~f:(make_cmt ~cm_kind:Cmi m)
```
Comment (rgrinberg) [inline src/dune_rules/install_rules.ml:190]: do we need these changes? They seem like unrelated refactoring to me.

File: src/dune_rules/lib_file_deps.ml [hunk 16]
```diff
@@ -1,22 +1,31 @@
 open Import
 
 module Group = struct
-  type t =
+  type ocaml =
     | Cmi
     | Cmx
+
+  type melange = Melange.Cm_kind.t
```
Comment (rgrinberg) [inline src/dune_rules/lib_file_deps.ml:8]: We don't need this alias I think.

File: src/dune_rules/lib_file_deps.ml [hunk 17]
```diff
@@ -1,22 +1,31 @@
...
+    | Ocaml Cmx -> Cm_kind.ext Cmx
+    | Melange Cmi -> Lib_mode.Cm_kind.ext (Melange Cmi)
+    | Melange Cmj -> Lib_mode.Cm_kind.ext (Melange Cmj)
     | Header -> Foreign_language.header_extension
 
   let obj_dir t obj_dir =
     match t with
-    | Cmi -> Obj_dir.public_cmi_dir obj_dir
-    | Cmx -> Obj_dir.native_dir obj_dir
+    | Ocaml Cmi | Melange Cmi -> Obj_dir.public_cmi_dir obj_dir
```
Comment (rgrinberg) [inline src/dune_rules/lib_file_deps.ml:26]: Shouldn't this be the Melange cmi dir?

File: src/dune_rules/context.mli [hunk 18]
```diff
@@ -105,8 +106,11 @@ val to_dyn_concise : t -> Dyn.t
 (** Compare the context names *)
 val compare : t -> t -> Ordering.t
 
+(** Return the OCaml compiler needed for the OCaml compilation mode *)
+val ocaml_compiler : t -> Mode.t -> Action.Prog.t
```
Comment (rgrinberg) [inline src/dune_rules/context.mli:110]: Do we need this function? I think we can just use `Context.compiler` everywhere.

File: src/dune_rules/cm_files.mli (hunk 11 above)
Comment (jchavarri) [reply src/dune_rules/cm_files.mli]: Nope, great catch, it was used for the cma hack, but I forgot to remove it in https://github.com/jchavarri/dune/pull/4/commits/7882f6ff3b30aea09008c208a4a3f67d33bb2db8. Removed in https://github.com/jchavarri/dune/pull/4/commits/129567ffa3a6cef42935ae34c21614c70cdd5f87.

File: src/dune_rules/cm_files.mli [hunk 19]
```diff
@@ -15,8 +15,12 @@ val make :
   -> unit
   -> t
 
+val filter_excluded_modules : t -> Module.t list -> Module.t list
```
[the remark this answers is not in the document]
Comment (jchavarri) [reply src/dune_rules/cm_files.mli]: Nope 😕  removed in https://github.com/jchavarri/dune/pull/4/commits/008cb504b51316f5a948ef56d35c6764c5a9e3fc.

File: src/dune_rules/dune_file.mli (hunk 12 above)
Comment (jchavarri) [reply src/dune_rules/dune_file.mli]: done in https://github.com/jchavarri/dune/pull/4/commits/99ef97ee7c78c79a0f47ba57362ead24b18d4356. I applied to another existing definition of `mode_conf` in `Mode_conf.Set`, hope it's fine.

File: src/dune_rules/dune_file.mli (hunk 13 above)
Comment (jchavarri) [reply src/dune_rules/dune_file.mli]: https://github.com/jchavarri/dune/pull/4/commits/99ef97ee7c78c79a0f47ba57362ead24b18d4356

File: src/dune_rules/melange.ml (hunk 14 above)
Comment (jchavarri) [reply src/dune_rules/melange.ml]: https://github.com/jchavarri/dune/pull/4/commits/d1c3a8953cdfc4a4600c84bd9148b1a4a0dc0a29

File: src/dune_rules/install_rules.ml (hunk 15 above)
Comment (jchavarri) [reply src/dune_rules/install_rules.ml]: good point, simplified in https://github.com/jchavarri/dune/pull/4/commits/ff572a3ea22c63104d5ff9952fd82c7a9a320b20.

File: src/dune_rules/lib_file_deps.ml (hunk 16 above)
Comment (jchavarri) [reply src/dune_rules/lib_file_deps.ml]: https://github.com/jchavarri/dune/pull/4/commits/a4c3b5661a8f479b58b2585f8a4e4248ba1d3110

File: src/dune_rules/lib_file_deps.ml (hunk 17 above)
Comment (jchavarri) [reply src/dune_rules/lib_file_deps.ml]: I think so. Still unclear how public cmi dirs will work with melange but for now using melange_dir for all melange kinds (https://github.com/jchavarri/dune/pull/4/commits/9f2477cec08fb14998533a38e1df0731be058051).

File: src/dune_rules/context.mli (hunk 18 above)
Comment (jchavarri) [reply src/dune_rules/context.mli]: Removed in https://github.com/jchavarri/dune/pull/4/commits/73108d0e600572cad907a45a9f86107dd3974f46.

File: src/dune_rules/melange.ml (hunk 14 above)
Comment (jchavarri) [reply src/dune_rules/melange.ml]: I thought we wanted to apply this change in all new modules, so replaced appearances in `Lib_mode` in https://github.com/jchavarri/dune/pull/4/commits/23c5b84c22e944f59ae338567a23230572527cd6.

**Comment by jchavarri** [comment]
Continues in https://github.com/ocaml/dune/pull/6238.
